[iOS] Make eager soft-linking of Data Detection frameworks non-fatal

brentfulgham opened the pull request:
#### ee46c752943bd6be0e433b8e7931efd5de456e77
<pre>
[iOS] Make eager soft-linking of Data Detection frameworks non-fatal
<a href="https://bugs.webkit.org/show_bug.cgi?id=255369">https://bugs.webkit.org/show_bug.cgi?id=255369</a>
&lt;rdar://107752677&gt;

Reviewed by David Kilzer.

Eager soft-linking of the Data Detection framework was added in Bug 241267.

Crash telemetry indicates that the soft-linking fails in some cases, possibly
when multiple code paths are trying to initialize the DataDetectors framework,
and one of those paths fails

Instead, we should consolidate the SoftLink code into a single implementation
file, and use that single code path in all cases.

* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/spi/ios/DataDetectorsUISPI.h: Moved some declarations
to DataDetectorsUISoftLink.
* Source/WebCore/PAL/pal/spi/ios/DataDetectorsUISoftLink.h: Added.
* Source/WebCore/PAL/pal/spi/ios/DataDetectorsUISoftLink.mm: Added.
* Source/WebCore/editing/cocoa/DataDetection.mm: Reference new SoftLink calls.
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm: Ditto.
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::softlinkDataDetectorsFrameworks): Ditto.

Canonical link: <a href="https://commits.webkit.org/263154@main">https://commits.webkit.org/263154@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d3bee3fce9c5af715e2434ed9a59e0acea248121

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/3775 "6 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/3861 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/3969 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/5205 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/4041 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/3747 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/3939 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/3866 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/3293 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/3820 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/4029 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/3397 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/5038 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/1533 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/3371 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/3859 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/3348 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/3430 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/4804 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/3825 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/3097 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/3349 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/3371 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/924 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/3390 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/3627 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

ddkilzer reviewed (approved):
r=me, but you need to fix the iOS build failures.

brentfulgham committed:
[iOS] Make eager soft-linking of Data Detection frameworks non-fatal
https://bugs.webkit.org/show_bug.cgi?id=255369
<rdar://107752677>

Reviewed by David Kilzer.

Eager soft-linking of the Data Detection framework was added in Bug 241267.

Crash telemetry indicates that the soft-linking fails in some cases, possibly
when multiple code paths are trying to initialize the DataDetectors framework,
and one of those paths fails

Instead, we should consolidate the SoftLink code into a single implementation
file, and use that single code path in all cases.

* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/spi/ios/DataDetectorsUISPI.h: Moved some declarations
to DataDetectorsUISoftLink.
* Source/WebCore/PAL/pal/spi/ios/DataDetectorsUISoftLink.h: Added.
* Source/WebCore/PAL/pal/spi/ios/DataDetectorsUISoftLink.mm: Added.
* Source/WebCore/editing/cocoa/DataDetection.mm: Reference new SoftLink calls.
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm: Ditto.
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::softlinkDataDetectorsFrameworks): Ditto.

Canonical link: https://commits.webkit.org/263154@main
